docs/ops/state/checkpoints.md

```diff
@@ -99,3 +99,296 @@ above).
+history of old checkpoints can grow indefinitely over time (like a changelog).
+
+At this point, it looks like we didn’t gain too much from incremental checkpoints because we are again trading between
+checkpointing overhead and recovery time. Fortunately, there are ways to improve on this naive approach to recovery. One
+simple and obvious way to restrict recovery time and the length of the checkpoint history is to write a full checkpoint
+from time to time. We can drop all checkpoints prior to the most recent full checkpoint, and the full checkpoint can
+serve as a new basis for future incremental checkpoints.
+
+Our actual implementation of incremental checkpoints in Flink is more involved and designed to address a number of
+different tradeoffs. Our incremental checkpointing restricts the size of the checkpoint history and therefore never
+needs take a full checkpoint to keep recovery efficiently! We present more detail about this in the next section, but
+the high level idea is to accept a small amount of redundant state writing to incrementally introduce
+merged/consolidated replacements for previous checkpoints. For now, you can think about Flink’s approach as stretching
+out and distributing the consolidation work over several incremental checkpoints, instead of doing it all at once in a
+full checkpoint. Every incremental checkpoint can contribute a share for consolidation. We also track when old
+checkpoints data becomes obsolete and then prune the checkpoint history over time.
+
+### Incremental Checkpoints in Flink
+
+In the previous section, we discussed that incremental checkpointing is mainly about recording all effective state
+modifications between checkpoints. This poses certain requirements on the underlying data structures in the state
+backend that manages the job’s state. It goes without saying that the data structure should always provide a decent
+read-write performance to keep state access swift. At the same time, for incremental checkpointing, the state backend
+must be able to efficiently detect and iterate state modifications since the previous checkpoint.
+
+One data structure that is very well-suited for this use case is the *log-structured-merge (LSM) tree* that is the core
+data structure in Flink’s RocksDB-based state backend. Without going into too much detail, the write path of RocksDB
+already roughly resembles a changelog with some pre-aggregation — which perfectly fits the needs of incremental
+checkpoints. On top of that, RocksDB also has a *compaction mechanism* can be regarded as an elaborated form of log
+compaction.
+
+#### RocksDB Snapshots as a Foundation
+
+In a nutshell, *RocksDB is a key-value store based on LSM trees*. The write path of RocksDB first collects all changes as
+key-value pairs in a mutable, in-memory buffer called memtable. Writes to the same key in a memtable can simply replace
```

Comment:
called a memtable. ... Once a memtable is full, ... After a memtable is written to disk, it becomes immutable and is then called a ...